packages/airnode-node/src/api/index.ts

```diff
@@ -0,0 +1,233 @@
+          sponsorAddress: sponsorAddress,
+          sponsorWalletAddress: sponsorWalletAddress,
+          endpointId: endpointId,
+          requestId: id,
+          chainId: chainId,
+          chainType: chain.type,
+          airnodeRrpAddress: chain.contracts.AirnodeRrp,
+        },
+      };
+    }
+  }
+}
+
+async function signResponseMessage(requestId: string, responseValue: string, config: Config) {
+  const masterHDNode = getMasterHDNode(config);
+  const airnodeWallet = ethers.Wallet.fromMnemonic(masterHDNode.mnemonic!.phrase);
+
+  return await airnodeWallet.signMessage(
+    ethers.utils.arrayify(
+      ethers.utils.keccak256(ethers.utils.solidityPack(['bytes32', 'bytes'], [requestId, responseValue || '0x']))
+    )
+  );
+}
+
+export interface CallApiPayload {
+  readonly config: Config;
+  readonly aggregatedApiCall: AggregatedApiCall;
+}
+
+function verifySponsorWallet(
+  config: Config,
+  aggregatedApiCall: RegularAggregatedApiCall
+): LogsData<ApiCallErrorResponse> | null {
+  const { sponsorAddress, sponsorWalletAddress, id } = aggregatedApiCall;
+  const hdNode = getMasterHDNode(config);
+  if (isValidSponsorWallet(hdNode, sponsorAddress, sponsorWalletAddress)) return null;
+
+  // TODO: Abstract this to a logging utils file
+  const message = `${RequestErrorMessage.SponsorWalletInvalid}, Request ID:${id}`;
+  const log = logger.pend('ERROR', message);
+  return [
+    [log],
+    {
+      success: false,
+      errorMessage: message,
+    },
+  ];
+}
+
+function verifyRequestId(
+  config_: Config,
+  aggregatedApiCall: RegularAggregatedApiCall
+): LogsData<ApiCallErrorResponse> | null {
+  if (isValidRequestId(aggregatedApiCall)) return null;
+
+  const message = `${RequestErrorMessage.RequestIdInvalid}. Request ID:${aggregatedApiCall.id}`;
+  const log = logger.pend('ERROR', message);
+  return [
+    [log],
+    {
+      success: false,
+      errorMessage: message,
+    },
+  ];
+}
+
+export function verifyTemplateId(
+  config_: Config,
+  aggregatedApiCall: RegularAggregatedApiCall
+): LogsData<ApiCallErrorResponse> | null {
+  const { templateId, template, id } = aggregatedApiCall;
+  if (!templateId) {
+    return null;
+  }
+
+  if (!template) {
+    const message = `Ignoring Request:${id} as the template could not be found for verification`;
+    const log = logger.pend('ERROR', message);
+    return [
+      [log],
+      {
+        success: false,
+        errorMessage: message,
+      },
+    ];
+  }
+
+  const expectedTemplateId = getExpectedTemplateId(template);
+  if (templateId === expectedTemplateId) return null;
+
+  const message = `Invalid template ID:${templateId} found for Request:${id}. Expected template ID:${expectedTemplateId}`;
+  const log = logger.pend('ERROR', message);
+  return [[log], { success: false, errorMessage: message }];
+}
+
+function verifyCallApiPayload(payload: CallApiPayload) {
+  const { config, aggregatedApiCall } = payload;
+  const verifications =
```

Comment:
I don't really agree. Currently, when doing an HTTP gateway request, the flow goes through 3 verification function calls just to figure out it should not do anything. I know it probably won't have an effect on performance but still.

If I look at the current code (line 153) I see that all requests go through 3 validations, which is technically not true, because there are exceptions for `testing-gateway` requests in the validation.
If I look at the proposed code, I immediately see that `testing-gateway` requests are treated a bit differently from normal requests.

But I agree that the current solution is better type-wise and it will have to be changed if verification will be added for `testing-gateway` requests.

I don't really feel strongly about this so I will happily revert it if you insist.

Reply:
I think it would be better.